Source/DafnyLanguageServer/Workspace/Notifications/VerificationDiagnosticsParams.cs

```diff
@@ -0,0 +1,664 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using System.Reflection.Metadata;
+using MediatR;
+using Microsoft.Boogie;
+using Microsoft.Dafny.LanguageServer.Language;
+using OmniSharp.Extensions.JsonRpc;
+using OmniSharp.Extensions.LanguageServer.Protocol;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+using Range = OmniSharp.Extensions.LanguageServer.Protocol.Models.Range;
+
+#pragma warning disable CS8618 // Non-nullable field must contain a non-null value when exiting constructor. Consider declaring as nullable.
+
+namespace Microsoft.Dafny.LanguageServer.Workspace.Notifications {
+  /// <summary>
+  /// DTO used to communicate the current compilation status to the LSP client.
+  /// </summary>
+  [Method(DafnyRequestNames.VerificationDiagnostics, Direction.ServerToClient)]
+  public class VerificationDiagnosticsParams : IRequest, IRequest<Unit> {
+    /// <summary>
+    /// Gets the URI of the document whose verification completed.
+    /// </summary>
+    public DocumentUri Uri { get; init; }
+
+    /// <summary>
+    /// Gets the version of the document.
+    /// </summary>
+    public int? Version { get; init; }
+
+    /// <summary>
+    /// Gets the same diagnostics as displayed in the diagnostics window
+    /// </summary>
+    public Container<Diagnostic> Diagnostics { get; init; }
+
+    /// <summary>
+    /// Returns a tree of diagnostics that can be used
+    /// First-level nodes are methods, or witness subset type checks
+    /// Second-level nodes are preconditions, postconditions, body verification status
+    /// Third level nodes are assertions inside functions 
+    /// </summary>
+    public NodeDiagnostic[] PerNodeDiagnostic { get; init; }
+
+    /// <summary>
+    /// The number of lines in the document
+    /// </summary>
+    public int LinesCount { get; init; }
+
+    public int NumberOfResolutionErrors { get; init; }
+
+    /// <summary>
+    /// Returns per-line real-time diagnostic
+    /// </summary>
+    public LineVerificationStatus[] PerLineDiagnostic { get; set; } = new LineVerificationStatus[] { };
+
+    public void RecomputePerLineDiagnostics() {
+      PerLineDiagnostic = RenderPerLineDiagnostics(this, PerNodeDiagnostic, LinesCount,
+        NumberOfResolutionErrors, Diagnostics);
+    }
```

Comment:
This is where the magic occurs. It works like a depth buffer: Every node is rendered on the same array of diagnostics for the gutter, but the priority determines what to really display, e.g. errors have a stronger priority than verified assertions, which have a stronger priority than error ranges.